json-rpc/src/stream_rpc/connection/client_connection.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    errors::JsonRpcError,
+    stream_rpc::{
+        connection::{ConnectionContext, StreamSender, Task},
+        counters,
+        json_rpc::{JsonRpcRequest, JsonRpcResponse, Method},
+        logging,
+        subscriptions::SubscriptionConfig,
+    },
+};
+use diem_infallible::Mutex;
+use diem_json_rpc_types::Id;
+use diem_logger::debug;
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+use storage_interface::DbReader;
+
+/// The `ClientConnection` is the interface between a transport, and subscriptions
+/// This will get cloned for every subscription, so lets keep it light :-)
+#[derive(Debug, Clone)]
+pub struct ClientConnection {
+    pub id: u64,
+    pub sender: StreamSender,
+    pub tasks: Arc<Mutex<HashMap<Id, Task>>>,
+    pub connection_context: Arc<ConnectionContext>,
+    pub config: Arc<SubscriptionConfig>,
+}
+
+impl ClientConnection {
+    pub fn new(
+        id: u64,
+        sender: StreamSender,
+        connection_context: ConnectionContext,
+        config: Arc<SubscriptionConfig>,
+    ) -> Self {
+        Self {
+            id,
+            sender,
+            tasks: Arc::new(Mutex::new(HashMap::new())),
+            connection_context: Arc::new(connection_context),
+            config,
+        }
+    }
+
+    /// When we disconnect a client, ensure we kill all of the asynchronous tasks (subscriptions)
+    pub fn disconnect(&self) {
+        let mut tasks = self.tasks.lock();
+        debug!("Aborting {} tasks for Client#{}", tasks.len(), &self.id);
+        tasks.iter().for_each(|(_id, task)| task.abort());
+        tasks.clear();
+    }
+
+    pub async fn send_raw(&self, message: String) -> anyhow::Result<()> {
+        if self.sender.is_closed() {
+            anyhow::bail!("Client#{} is closed", self.id);
+        }
+        match self.sender.send(Ok(message)).await {
+            Ok(_) => Ok(()),
+            Err(e) => match e.0 {
+                Ok(_) => Ok(()),
+                Err(e) => Err(e),
+            },
+        }
+    }
+
+    pub async fn send_response(&self, message: JsonRpcResponse) -> anyhow::Result<()> {
+        if let Ok(response) = serde_json::to_string(&message) {
+            return self.send_raw(response).await;
+        }
+        anyhow::bail!("Could not convert {:?} to string", message)
```

Comment:
as this is a library I would recommend more formal error types since anyhow's are basically useless except as a generic error.

Reply:
do we try again, do we just fail, ...